store/pruning.go

```diff
@@ -0,0 +1,75 @@
+package store
+
+import "fmt"
+
+// PruneBlocks removes block up to (but not including) a height. It returns number of blocks pruned.
+func (s *DefaultStore) PruneBlocks(heightInt int64) (uint64, error) {
+	if heightInt <= 0 {
+		return 0, fmt.Errorf("height must be greater than 0")
+	}
+
+	//TODO: validate the height is > committed/synced height
+	height := uint64(heightInt)
+	if height > s.Height() {
+		return 0, fmt.Errorf("cannot prune beyond the latest height %v", s.height)
+	}
+	base := s.Base()
+	if height < base {
+		return 0, fmt.Errorf("cannot prune to height %v, it is lower than base height %v",
+			height, base)
+	}
+
+	pruned := uint64(0)
+	batch := s.db.NewBatch()
+	defer batch.Discard()
+
+	flush := func(batch Batch, base uint64) error {
+		s.SetBase(base)
+		err := batch.Commit()
+		if err != nil {
+			return fmt.Errorf("failed to prune up to height %v: %w", base, err)
+		}
+		return nil
+	}
+
+	for h := base; h < height; h++ {
+		hash, err := s.loadHashFromIndex(h)
+		if err != nil {
+			continue
+		}
+		if err := batch.Delete(getBlockKey(hash)); err != nil {
+			return 0, err
+		}
+		if err := batch.Delete(getCommitKey(hash)); err != nil {
+			return 0, err
+		}
+		if err := batch.Delete(getIndexKey(h)); err != nil {
+			return 0, err
+		}
+		if err := batch.Delete(getResponsesKey(h)); err != nil {
+			return 0, err
+		}
+		if err := batch.Delete(getValidatorsKey(h)); err != nil {
+			return 0, err
+		}
+
+		pruned++
+
+		// flush every 1000 blocks to avoid batches becoming too large
+		if pruned%1000 == 0 && pruned > 0 {
+			err := flush(batch, h)
+			if err != nil {
+				return 0, err
+			}
+			//FIXME: I think it mess with the defer above as it overwrites the batch variable
```

Comment:
don't think there is a problem here. it will simply defer both functions. it doesn't override the previous defer.